### ISSUE TITLE
No item named 'ppt/drawings/NULL' in the archive

### ISSUE THREAD

**Issue by joswr1ght** [scanny/python-pptx#206]
When I open a particular PPTX file (which opens without error or complaint in multiple versions/platforms of PowerPoint) I get the following exception:

```
$ python
Python 2.7.10 (v2.7.10:15c95b7d81dc, May 23 2015, 09:33:12)
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pptx import Presentation
>>> prs = Presentation("MyPresentation.pptx")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/api.py", line 26, in __init__
    self._package = Package.open(pkg_file)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/package.py", line 44, in open
    return super(Package, cls).open(pkg_file)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/opc/package.py", line 122, in open
    pkg_reader = PackageReader.from_file(pkg_file)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/opc/pkgreader.py", line 36, in from_file
    phys_reader, pkg_srels, content_types
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/opc/pkgreader.py", line 69, in _load_serialized_parts
    for partname, blob, srels in part_walker:
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/opc/pkgreader.py", line 104, in _walk_phys_parts
    phys_reader, part_srels, visited_partnames):
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/opc/pkgreader.py", line 104, in _walk_phys_parts
    phys_reader, part_srels, visited_partnames):
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/opc/pkgreader.py", line 104, in _walk_phys_parts
    phys_reader, part_srels, visited_partnames):
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/opc/pkgreader.py", line 101, in _walk_phys_parts
    blob = phys_reader.blob_for(partname)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pptx/opc/phys_pkg.py", line 109, in blob_for
    return self._zipf.read(pack_uri.membername)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/zipfile.py", line 935, in read
    return self.open(name, "r", pwd).read()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/zipfile.py", line 961, in open
    zinfo = self.getinfo(name)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/zipfile.py", line 909, in getinfo
    'There is no item named %r in the archive' % name)
KeyError: "There is no item named 'ppt/drawings/NULL' in the archive"
```

I'm using Pillow-3.1.1 XlsxWriter-0.8.4 lxml-3.6.0 python-pptx-0.5.8 on Python 2.7.10 on OS X.  Thoughts on this error?
Thanks!
-Josh

**Comment by scanny** [comment]
This problem came up quite a while back with a .pptx that was generated (or modified, I can't remember which) with I believe it was a Java package (not MS PowerPoint). That package took a shortcut in deleting images or something, just setting them to "NULL" instead of cleaning up after itself.

I think if you open the file in PowerPoint and then save, this might go away. Without knowing your use case of course I can't say whether this is feasible or not.

If not, let me know more about your situation and I may be able to advise. You could probably use some of the python-pptx internals to clean up the file as a preprocessing step if you're going to get a lot like this. Or maybe a patch somewhere will do the trick and we can incorporate it into a future version.

**Comment by joswr1ght** [comment]
I narrowed the problem down to a single slide in a larger pptx file, attached.  This slide was created in PowerPoint, and I saved it several times but the python-pptx problem persists.  I even saved it as a "ppt" file, then opened that and saved it as a "pptx" file, but the problem remains.

Is there something that can be done in the library to return a ```Presentation``` instance despite this error?

Thanks!
[new.pptx](https://github.com/scanny/python-pptx/files/198824/new.pptx)

**Comment by pablonieto0981** [comment]
Hi scanny, python pptx is a great package. Thanks!

I just worked on this problem today and found, after some time going up and down, the following solution:

In the **pkgreader.py** module, I changed this...

    def load_from_xml(baseURI, rels_item_xml):
        """
        Return |_SerializedRelationshipCollection| instance loaded with the
        relationships contained in *rels_item_xml*. Returns an empty
        collection if *rels_item_xml* is |None|.
        """
        srels = _SerializedRelationshipCollection()
        if rels_item_xml is not None:
            rels_elm = parse_xml(rels_item_xml)
            for rel_elm in rels_elm.relationship_lst:
                srels._srels.append(_SerializedRelationship(baseURI, rel_elm))
        return srels

...for this...

    def load_from_xml(baseURI, rels_item_xml):
        """
        Return |_SerializedRelationshipCollection| instance loaded with the
        relationships contained in *rels_item_xml*. Returns an empty
        collection if *rels_item_xml* is |None|.
        """
        srels = _SerializedRelationshipCollection()
        if rels_item_xml is not None:
            **if 'NULL' in str(rels_item_xml):
                print('NULL found and skiped')
                pass
            else:
                rels_elm = parse_xml(rels_item_xml)
                for rel_elm in rels_elm.relationship_lst:
                    srels._srels.append(_SerializedRelationship(baseURI, rel_elm))**
        return srels

I hope this helps!

**Comment by otiai10** [comment]
Hi @pablonieto0981, why don't you send pull-request with it? It's gonna be appreciated!
Hi @scanny, your package saves my days many times. Just to say thank you!

**Comment by mhdsRahnama** [comment]
I have similar problem in reading .docx file with python-docx.
the error is:
KeyError: "There is no item named 'word/NULL' in the archive"

**Comment by blue-chip** [comment]
I had a similar issue but with `ppt/slideLayouts/NULL`. After a lot of thinkering, I managed to solve this by manually deleting all the corrupted Slide Layouts in the Slide Master of my pptx.

First, I unzipped the pptx and opened the folder in VS Code. There, I searched for all NULL occurrences which gave me a list of `.rels` files in the `ppt/slideLayouts/_rels/` directory. For each file, I looked up it's equivalent `.xml` file in `ppt/slideLayouts/` which allowed me to get the slide layout name used in the PowerPoint UI. I then looked it up in the Slide Master in PowerPoint and deleted it. In my case, I could do that because those corrupted layouts weren't even used.

Almost all steps are manual but at least it solved the issue in my case.

**Comment by victai** [comment]
I wrote a script that goes through all .rels files, find items with target="NULL", and remove those items from the corresponding .xml file. 
ex. If there's an item='NULL' in `ppt/slides/_rels/slide1.xml.rels`, I would get the Id of the item, then remove the object with the same Id in `ppt/slides/slide1.xml`

I'm not sure if it covers all possibilities, and I have not tested extensively, but hope this helps.

```
import os
import zipfile
import tempfile
from lxml import etree

def remove_NULL_from_rels(data):
    ET = etree.fromstring(data)
    rels_to_remove = []
    for node in ET.iter():
        if node.attrib.get('Target') == 'NULL':
            rels_to_remove.append(node.attrib['Id'])
            node.getparent().remove(node)
    return etree.tostring(ET, xml_declaration=True), rels_to_remove

def remove_rels_from_slide(data, rels):
    ET = etree.fromstring(data)
    for node in ET.iter():
        for rId in rels:
            if rId in node.attrib.values():
                node.getparent().remove(node)
    return etree.tostring(ET, xml_declaration=True)

def fixCorruptedPPTX(zipname):
    # generate a temp file
    tmpfd, tmpname = tempfile.mkstemp(dir=os.path.dirname(zipname))
    os.close(tmpfd)

    # create a temp copy of the archive without filename            
    with zipfile.ZipFile(zipname, 'r') as zin:
        with zipfile.ZipFile(tmpname, 'w') as zout:
            zout.comment = zin.comment # preserve the comment
            rels_xml_mapping = {}
            modified_xml = {}
            modified_rels = {}
            for item in zin.infolist():
                if item.filename.endswith('.rels'):
                    rels_file = item.filename
                    rels_dir, rels_name = os.path.split(rels_file)
                    rels_data = zin.read(rels_file)
                    rels_data, rels_to_remove = remove_NULL_from_rels(rels_data)
                    xml_file = os.path.join(os.path.split(rels_dir)[0], rels_name.rsplit('.', 1)[0])
                    if len(rels_to_remove) > 0 and xml_file in zin.namelist():
                        xml_data = zin.read(xml_file)
                        xml_data = remove_rels_from_slide(xml_data, rels_to_remove)
                        modified_xml[xml_file] = xml_data
                        modified_rels[rels_file] = rels_data
                
            for item in zin.infolist():
                if item.filename not in modified_xml and item.filename not in modified_rels:
                    zout.writestr(item, zin.read(item.filename))
                    
    # replace with the temp archive
    os.remove(zipname)
    os.rename(tmpname, zipname)

    # now add filename with its new data
    with zipfile.ZipFile(zipname, mode='a', compression=zipfile.ZIP_DEFLATED) as zf:
        for rels_file, rels_data in modified_rels.items():
            zf.writestr(rels_file, rels_data)
        for xml_file, xml_data in modified_xml.items():
            zf.writestr(xml_file, xml_data)

fixCorruptedPPTX(pptx_file)
```

**Comment by monstarnn** [comment]
@victai thank you, it works perfect!

**Comment by scanny** [comment]
This should be fixed in the upcoming release `0.6.20` due out today or tomorrow. Any relationship "pointing" to a package part that cannot be found (e.g. "NULL") is now ignored. If the file is truly corrupted and a part is in fact missing, this isn't going to help, but the common-ish case where a neglectful client has "removed" relationships simply by marking them "NULL" no longer raises an error on load.
